### PR TITLE
P2338R4 Freestanding Library: Character primitives and the C library

### LINKED ISSUE #6298: [2023-06 LWG Motion 12] P2338R4 Freestanding Library: Character primitives and the C library
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2338r4.html

### PR DESCRIPTION
**Pull request by jensmaurer**
- sort table [tab:headers.cpp.fs] by section number

Fixes #6298
Fixes cplusplus/papers#1017

### CODE REVIEW

File: source/lib-intro.tex [hunk 1]
```diff
@@ -948,6 +948,21 @@
 \end{codeblock}
 \end{example}
 
+\pnum
+\indextext{function!freestanding deleted}%
+Function declarations and function template declarations
+followed by a comment that include \textit{freestanding-deleted} are
+freestanding deleted functions.
```
Comment (JohelEGP) [inline source/lib-intro.tex:955]:
```suggestion
Function declarations and function template declarations
followed by a comment that include \textit{freestanding-deleted} are
\defnadjx{freestanding deleted}{functions}{function}.
```
```suggestion
\indextext{function!freestanding deleted}%
Function declarations and function template declarations
followed by a comment that include \textit{freestanding-deleted} are
\textit{freestanding deleted functions}.
```

File: source/support.tex [hunk 2]
```diff
@@ -209,30 +209,30 @@
 
   // \ref{alg.c.library}, C standard library algorithms
   void* bsearch(const void* key, const void* base, size_t nmemb, size_t size,
-                @\placeholder{c-compare-pred}@*@\itcorr[-1]@ compar);
+                @\placeholder{c-compare-pred}@*@\itcorr[-1]@ compar);                                // freestanding
   void* bsearch(const void* key, const void* base, size_t nmemb, size_t size,
-                @\placeholder{compare-pred}@*@\itcorr[-1]@ compar);
-  void qsort(void* base, size_t nmemb, size_t size, @\placeholder{c-compare-pred}@*@\itcorr[-1]@ compar);
-  void qsort(void* base, size_t nmemb, size_t size, @\placeholder{compare-pred}@*@\itcorr[-1]@ compar);
+                @\placeholder{compare-pred}@*@\itcorr[-1]@ compar);                                  // freestanding
```
Comment (JohelEGP) [inline source/support.tex:214]: I think it's preferable to have these comments
be on the same line as the declaration's identifier they apply to.

File: source/utilities.tex [hunk 3]
```diff
@@ -14188,36 +14188,40 @@
...
 
-  to_chars_result to_chars(char* first, char* last, @\placeholder{floating-point-type}@ value);
-  to_chars_result to_chars(char* first, char* last, @\placeholder{floating-point-type}@ value, chars_format fmt);
+  to_chars_result to_chars(char* first, char* last,
+                           @\placeholder{floating-point-type}@ value);            // freestanding-deleted
   to_chars_result to_chars(char* first, char* last, @\placeholder{floating-point-type}@ value,
-                           chars_format fmt, int precision);
+                           chars_format fmt);                           // freestanding-deleted
+  to_chars_result to_chars(char* first, char* last, @\placeholder{floating-point-type}@ value,
+                           chars_format fmt, int precision);            // freestanding-deleted
```
Comment (JohelEGP) [inline source/utilities.tex:14207]: I think it's preferable to have these comments
be on the same line as the declaration's identifier they apply to.

File: source/utilities.tex [hunk 4]
```diff
@@ -14188,36 +14188,40 @@
...
     friend bool operator==(const from_chars_result&, const from_chars_result&) = default;
   };
 
   constexpr from_chars_result from_chars(const char* first, const char* last,
-                                         @\placeholder{integer-type}@& value, int base = 10);
+                                         @\placeholder{integer-type}@& value, int base = 10);     // freestanding
 
   from_chars_result from_chars(const char* first, const char* last, @\placeholder{floating-point-type}@& value,
-                               chars_format fmt = chars_format::general);
+                               chars_format fmt = chars_format::general);   // freestanding-deleted
```
Comment (JohelEGP) [inline source/utilities.tex:14224]: I think it's preferable to have these comments
be on the same line as the declaration's identifier they apply to.

**Review by JohelEGP — CHANGES_REQUESTED** [review]
The application of the following instruction is missing:
> Please add the following feature test macros to [[version.syn]](https://wg21.link/version.syn#2):

Note that such application introduces merge conflicts with #6355, as pointed out:
> This paper also assumes that [LWG3753](https://cplusplus.github.io/LWG/issue3753) and [P2198](https://wg21.link/P2198) have been accepted and applied.

File: source/support.tex (hunk 2 above)
Comment (jensmaurer) [reply source/support.tex]: I think I've fixed this.

File: source/utilities.tex (hunk 3 above)
Comment (jensmaurer) [reply source/utilities.tex]: I think I've fixed this.

File: source/utilities.tex (hunk 4 above)
Comment (jensmaurer) [reply source/utilities.tex]: I think I've fixed this.

**Review by JohelEGP — APPROVED** [review]

File: source/strings.tex [hunk 5]
```diff
@@ -5515,38 +5515,38 @@
...
+  char* strchr(char* s, int c);                                         // freestanding; see \ref{library.c}
+  size_t strcspn(const char* s1, const char* s2);                       // freestanding
+  const char* strpbrk(const char* s1, const char* s2);                  // freestanding; see \ref{library.c}
+  char* strpbrk(char* s1, const char* s2);                              // freestanding; see \ref{library.c}
+  const char* strrchr(const char* s, int c);                            // freestanding; see \ref{library.c}
+  char* strrchr(char* s, int c);                                        // freestanding; see \ref{library.c}
+  size_t strspn(const char* s1, const char* s2);                        // freestanding
+  const char* strstr(const char* s1, const char* s2);                   // freestanding; see \ref{library.c}
+  char* strstr(char* s1, const char* s2);                               // freestanding; see \ref{library.c}
+  char* strtok(char* s1, const char* s2);                               // freestanding
```
Comment (frederick-vs-ja) [inline source/strings.tex:5543]: `strtok` is now removed from freestanding library contents of C23 (see [this document](https://docs.google.com/document/d/1S-rgdTiqQ-vQkmxnwZm9JxLBFG2_Wr-3/edit)). Perhaps we need an LWG issue to strike `strtok` and `wcstok`.

Comment (jensmaurer) [reply source/strings.tex:5543]: This is being handled by a future paper.

**Pull request merged**